org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/hover/AbstractProblemHover.java

```diff
@@ -67,38 +68,81 @@ protected IAnnotationModel getAnnotationModel() {
 		return sourceViewer.getAnnotationModel();
 	}
 
+	/**
+	 * Find all handled annotations that include the given offset. If there are multiple annotations,
+	 * those that start on the given line number are put first in the list, then those that start one
+	 * line before, etc.
+	 */
 	public List<Annotation> getAnnotations(final int lineNumber, final int offset) {
 		if (getAnnotationModel() == null) {
 			return Collections.emptyList();
 		}
-		final Iterator<?> iterator = getAnnotationModel().getAnnotationIterator();
+		final Iterator<Annotation> iterator = getAnnotationModel().getAnnotationIterator();
 		List<Annotation> result = Lists.newArrayList();
+		List<Integer> lineDiffs = Lists.newArrayList();
 		while (iterator.hasNext()) {
-			final Annotation annotation = (Annotation) iterator.next();
-			if (isHandled(annotation)) {
+			final Annotation annotation = iterator.next();
+			if (isHandled(annotation) && !isLineDiffInfo(annotation)) {
 				Position position = getAnnotationModel().getPosition(annotation);
 				if (position != null) {
 					final int start = position.getOffset();
 					final int end = start + position.getLength();
 
-					if (offset > 0 && !(start <= offset && offset <= end)) {
-						continue;
-					}
-					try {
-						if (lineNumber != getDocument().getLineOfOffset(start)) {
-							continue;
+					if (offset < 0 || start <= offset && offset <= end) {
+						try {
+							int lineDiff = Math.abs(lineNumber - getDocument().getLineOfOffset(start));
+							if (lineDiff == 0 || offset >= 0) {
+								// Insert the annotation according to its line number difference
+								int index = lineDiffs.size();
+								while (index > 0 && lineDiffs.get(index - 1) > lineDiff) {
+									index--;
+								}
+								result.add(index, annotation);
+								lineDiffs.add(index, lineDiff);
+							}
+						} catch (Exception x) {
+							// Skip this annotation
 						}
-					} catch (final Exception x) {
-						continue;
-					}
-					if (!isLineDiffInfo(annotation)) {
-						result.add(annotation);
 					}
 				}
 			}
 		}
 		return result;
 	}
+	
+	/**
+	 * Sort the given annotations such that the resulting list contains first errors, then warnings, and then the rest.
+	 */
+	protected List<Annotation> sortBySeverity(List<Annotation> annotations) {
+		if (annotations == null)
+			return Collections.emptyList();
+		if (annotations.size() <= 1)
+			return annotations;
+		List<Annotation> sorted = Lists.newArrayList(annotations);
+		Collections.sort(sorted, new Comparator<Annotation>() {
+			@Override
+			public int compare(Annotation a1, Annotation a2) {
+				boolean isError1 = markerAnnotationAccess.isSubtype(a1.getType(), "org.eclipse.ui.workbench.texteditor.error");
```

Comment:
Can we extract two methods `isError`, `isWarning` probably also `isInfo` and use these is `isHandled`, too? This would avoid the duplicated string constants.

Reply:
Using Boolean.compare(..) could further simplify the code, e.g. Comparator.comparing(this::isError, Boolean::compare).thenComparing(this::isWarning, ..)